pkg/iam/iam.go

```diff
@@ -0,0 +1,327 @@
+	}
+	if config.ClusterName == "" {
+		return nil, errors.New("cannot create IAMService with empty ClusterName")
+	}
+	if config.IAMRoleName == "" {
+		return nil, errors.New("cannot create IAMService with empty IAMRoleName")
+	}
+	if config.Log == nil {
+		return nil, errors.New("cannot create IAMService with Log equal to nil")
+	}
+	if !(config.RoleType == ControlPlaneRole || config.RoleType == NodesRole) {
+		return nil, fmt.Errorf("cannot create IAMService with invalid RoleType '%s'", config.RoleType)
+	}
+	client := awsiam.New(config.AWSSession)
+
+	l := config.Log.WithValues("clusterName", config.ClusterName, "iam-role", config.RoleType)
+
+	s := &IAMService{
+		clusterName: config.ClusterName,
+		iamClient:   client,
+		iamRoleName: config.IAMRoleName,
+		log:         l,
+		roleType:    config.RoleType,
+		region:      client.SigningRegion,
+	}
+
+	return s, nil
+}
+
+func (s *IAMService) Reconcile() error {
+	s.log.Info("reconciling IAM role")
+
+	err := s.createMainRole()
+	if err != nil {
+		return err
+	}
+
+	// we only attach the inline policy to a role that is owned (and was created) by iam controller
+	owned, err := isOwnedByIAMController(s.iamRoleName, s.iamClient)
+	if err != nil {
+		s.log.Error(err, "Failed to fetch IAM Role")
+		return err
+	}
+	// check if the policy is created by this controller, if its not than we skip adding inline policy
+	if !owned {
+		s.log.Info("IAM role is not owned by IAM controller, skipping adding inline policy")
+	} else {
+		err = s.attachInlinePolicy()
+		if err != nil {
+			return err
+		}
+	}
+
+	s.log.Info("finished reconciling IAM role")
+	return nil
+}
+
+// createMainRole will create the main IAM role that will be attached to EC2 instances
+func (s *IAMService) createMainRole() error {
+	i := &awsiam.GetRoleInput{
+		RoleName: aws.String(s.iamRoleName),
+	}
+
+	_, err := s.iamClient.GetRole(i)
+
+	// create new IAMRole if it does not exists yet
+	if IsNotFound(err) {
+		assumeRolePolicyDocument, err := generateAssumeRolePolicyDocument(s.region)
+		if err != nil {
+			s.log.Error(err, "failed to generate assume policy document from template for IAM role")
+			return err
+		}
+
+		tags := []*awsiam.Tag{
+			{
+				Key:   aws.String(IAMControllerOwnedTag),
+				Value: aws.String(""),
+			},
+			{
+				Key:   aws.String(fmt.Sprintf(ClusterIDTag, s.clusterName)),
+				Value: aws.String("owned"),
+			},
+		}
+
+		i := &awsiam.CreateRoleInput{
+			RoleName:                 aws.String(s.iamRoleName),
+			AssumeRolePolicyDocument: aws.String(assumeRolePolicyDocument),
+			Tags:                     tags,
+		}
+
+		_, err = s.iamClient.CreateRole(i)
+		if err != nil {
+			s.log.Error(err, "failed to create main IAM Role")
+			return err
+		}
+
+		s.log.Info("successfully created a new main IAM role")
+	} else if err != nil {
+		s.log.Error(err, "Failed to fetch IAM Role")
+		return err
+	} else {
+		s.log.Info("IAM Role already exists, skipping creation")
```

Comment:
We could add the `clusterName` in case we need to debug, should be easier to grep

Reply:
the logger that is passed to the IAM service already has it 
https://github.com/giantswarm/capa-iam-controller/pull/11/files#diff-ee67ba6a4ebbae43367c7425da9e9a76aadf39445a0cd224aaba5810381711c5R63